File: android/lib/runtime/runtime_snapshot.dart
enum RuntimePhase {
  idle,
  starting,
  preparingHost,
  connecting,
  handshaking,
  applyingPolicy,
  connected,
  reconnecting,
  stopping,
  failed,
  unknown;

  static RuntimePhase parse(String value) {
    switch (value) {
      case 'idle':
        return RuntimePhase.idle;
      case 'starting':
        return RuntimePhase.starting;
      case 'preparing_host':
        return RuntimePhase.preparingHost;
      case 'connecting':
        return RuntimePhase.connecting;
      case 'handshaking':
        return RuntimePhase.handshaking;
      case 'applying_policy':
        return RuntimePhase.applyingPolicy;
      case 'connected':
        return RuntimePhase.connected;
      case 'reconnecting':
        return RuntimePhase.reconnecting;
      case 'stopping':
        return RuntimePhase.stopping;
      case 'failed':
        return RuntimePhase.failed;
      case 'unknown':
        return RuntimePhase.unknown;
      default:
        throw FormatException('Unknown runtime phase: $value');
    }
  }
}

enum P2PState {
  disabled('disabled', 'Disabled'),
  unavailable('unavailable', 'Unavailable'),
  relay('relay', 'Relay'),
  eligible('eligible', 'Eligible'),
  probing('probing', 'Probing'),
  direct('direct', 'Direct'),
  suspect('suspect', 'Suspect'),
  fallingBack('falling_back', 'Falling back'),
  failed('failed', 'Failed');

  const P2PState(this.wireName, this.displayName);

  final String wireName;
  final String displayName;

  static P2PState parse(String value) {
    for (final state in values) {
      if (state.wireName == value) return state;
    }
    return P2PState.unavailable;
  }
}

class RuntimeErrorSnapshot {
  const RuntimeErrorSnapshot({
    this.code = 0,
    this.severity = '',
    this.retryable = false,
    this.userMessageKey = '',
    this.diagnosticDetail = '',
  });

  final int code;
  final String severity;
  final bool retryable;
  final String userMessageKey;
  final String diagnosticDetail;

  factory RuntimeErrorSnapshot.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const RuntimeErrorSnapshot();
    return RuntimeErrorSnapshot(
      code: json['code'] as int? ?? 0,
      severity: json['severity'] as String? ?? '',
      retryable: json['retryable'] as bool? ?? false,
      userMessageKey: json['user_message_key'] as String? ?? '',
      diagnosticDetail: json['diagnostic_detail'] as String? ?? '',
    );
  }
}

class RuntimeTrafficSnapshot {
  const RuntimeTrafficSnapshot({
    this.rxBytes = 0,
    this.txBytes = 0,
  });

  final int rxBytes;
  final int txBytes;

  factory RuntimeTrafficSnapshot.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const RuntimeTrafficSnapshot();
    return RuntimeTrafficSnapshot(
      rxBytes: json['rx_bytes'] as int? ?? 0,
      txBytes: json['tx_bytes'] as int? ?? 0,
    );
  }
}

class RuntimeSnapshot {
  static const bundledCapabilities = <String>[
    'mux.compat',
    'mux.flow',
    'mux.balance',
    'mux.stripe',
  ];

  const RuntimeSnapshot({
    required this.generation,
    required this.monotonicMs,
    required this.phase,
    this.role = '',
    this.server = '',
    this.transport = '',
    this.capabilities = const <String>[],
    this.requestedMuxMode = '',
    this.effectiveMuxMode = '',
    this.muxReceiverOrdering = '',
    this.muxActiveLinks = 0,
    this.muxFallbackReason = '',
    this.p2pState = P2PState.disabled,
    this.traffic = const RuntimeTrafficSnapshot(),
    this.connectedMonotonicMs = 0,
    this.lastError = const RuntimeErrorSnapshot(),
  });

  static const int schemaVersion = 1;

  final int generation;
  final int monotonicMs;
  final RuntimePhase phase;
  final String role;
  final String server;
  final String transport;
  final List<String> capabilities;
  final String requestedMuxMode;
  final String effectiveMuxMode;
  final String muxReceiverOrdering;
  final int muxActiveLinks;
  final String muxFallbackReason;
  final P2PState p2pState;
  String get effectivePath => p2pState == P2PState.direct ? 'direct' : 'relay';
  final RuntimeTrafficSnapshot traffic;

  /// `monotonic_ms` at which the session entered `connected`, or 0 when it is
  /// not connected. Elapsed time is `monotonicMs - connectedMonotonicMs`, so it
  /// stays correct across a UI process restart.
  final int connectedMonotonicMs;
  final RuntimeErrorSnapshot lastError;

  factory RuntimeSnapshot.fromJson(Map<String, dynamic> json) {
    final version = json['schema_version'];
    if (version != schemaVersion) {
      throw FormatException('Unsupported runtime schema version: $version');
    }
    final generation = json['generation'];
    if (generation is! int || generation < 0) {
      throw const FormatException('Runtime generation is required');
    }
    final monotonicMs = json['monotonic_ms'];
    if (monotonicMs is! int || monotonicMs < 0) {
      throw const FormatException('Runtime monotonic_ms is required');
    }
    final phaseValue = json['phase'];
    if (phaseValue is! String) {
      throw const FormatException('Runtime phase is required');
    }

    return RuntimeSnapshot(
      generation: generation,
      monotonicMs: monotonicMs,
      phase: RuntimePhase.parse(phaseValue),
      role: json['role'] as String? ?? '',
      server: json['server'] as String? ?? '',
      transport: json['transport'] as String? ?? '',
      capabilities: json.containsKey('capabilities')
          ? (json['capabilities'] as List<dynamic>? ?? const <dynamic>[])
              .whereType<String>()
              .toList(growable: false)
          : bundledCapabilities,
      requestedMuxMode: json['requested_mux_mode'] as String? ?? '',
      effectiveMuxMode: json['effective_mux_mode'] as String? ?? '',
      muxReceiverOrdering: json['mux_receiver_ordering'] as String? ?? '',
      muxActiveLinks: json['mux_active_links'] as int? ?? 0,
      muxFallbackReason: json['mux_fallback_reason'] as String? ?? '',
      p2pState: P2PState.parse(json['p2p_state'] as String? ?? 'disabled'),
      traffic: RuntimeTrafficSnapshot.fromJson(
        json['traffic'] as Map<String, dynamic>?,
      ),
      connectedMonotonicMs: json['connected_monotonic_ms'] as int? ?? 0,
      lastError: RuntimeErrorSnapshot.fromJson(
        json['last_error'] as Map<String, dynamic>?,
      ),
    );
  }

  List<String> availableMuxModes({bool experimental = false}) {
    const modes = <String>['compat', 'flow', 'balance', 'stripe'];
    return modes
        .where((mode) => capabilities.contains('mux.$mode'))
        .where((mode) => mode != 'stripe' || experimental)
        .toList(growable: false);
  }

  String get effectiveMuxDisplayName =>
      effectiveMuxMode == 'compat' ? 'Compatibility mode' : effectiveMuxMode;

  List<String> get muxDiagnosticLines => <String>[
        if (requestedMuxMode.isNotEmpty) 'Requested VMUX: $requestedMuxMode',
        if (effectiveMuxMode.isNotEmpty)
          'Effective VMUX: $effectiveMuxDisplayName',
        if (muxFallbackReason.isNotEmpty)
          'Fallback reason: $muxFallbackReason',
      ];

  String get effectivePathDisplayName =>
      effectivePath == 'direct' ? 'Direct' : 'Relay';

  List<String> get p2pDiagnosticLines => <String>[
        'P2P: ${p2pState.displayName}',
        'Effective path: $effectivePathDisplayName',
      ];
}

File: .tmp-v215-src/ios/App/OpenPPP2/Runtime/RuntimeSnapshot.swift
import Foundation

public enum RuntimePhase: String, Codable, Sendable {
    case idle
    case starting
    case preparingHost = "preparing_host"
    case connecting
    case handshaking
    case applyingPolicy = "applying_policy"
    case connected
    case reconnecting
    case stopping
    case failed
    case unknown
}

public enum P2PState: String, CaseIterable, Sendable {
    case disabled
    case unavailable
    case relay
    case eligible
    case probing
    case direct
    case suspect
    case fallingBack = "falling_back"
    case failed

    public static func parse(_ value: String) -> P2PState {
        P2PState(rawValue: value) ?? .unavailable
    }

    public var displayName: String {
        switch self {
        case .disabled: return "Disabled"
        case .unavailable: return "Unavailable"
        case .relay: return "Relay"
        case .eligible: return "Eligible"
        case .probing: return "Probing"
        case .direct: return "Direct"
        case .suspect: return "Suspect"
        case .fallingBack: return "Falling back"
        case .failed: return "Failed"
        }
    }
}

public struct RuntimeErrorSnapshot: Codable, Equatable, Sendable {
    public var code: UInt32
    public var severity: String
    public var retryable: Bool
    public var userMessageKey: String
    public var diagnosticDetail: String

    public init(
        code: UInt32 = 0,
        severity: String = "",
        retryable: Bool = false,
        userMessageKey: String = "",
        diagnosticDetail: String = ""
    ) {
        self.code = code
        self.severity = severity
        self.retryable = retryable
        self.userMessageKey = userMessageKey
        self.diagnosticDetail = diagnosticDetail
    }

    enum CodingKeys: String, CodingKey {
        case code
        case severity
        case retryable
        case userMessageKey = "user_message_key"
        case diagnosticDetail = "diagnostic_detail"
    }
}

public struct RuntimeTrafficSnapshot: Codable, Equatable, Sendable {
    public var rxBytes: UInt64
    public var txBytes: UInt64

    public init(rxBytes: UInt64 = 0, txBytes: UInt64 = 0) {
        self.rxBytes = rxBytes
        self.txBytes = txBytes
    }

    enum CodingKeys: String, CodingKey {
        case rxBytes = "rx_bytes"
        case txBytes = "tx_bytes"
    }
}

public struct RuntimeSnapshot: Codable, Equatable, Sendable {
    public static let schemaVersion: UInt32 = 1
    public static let bundledCapabilities = [
        "mux.compat", "mux.flow", "mux.balance", "mux.stripe"
    ]

    public var generation: UInt64
    public var monotonicMs: UInt64
    public var phase: RuntimePhase
    public var role: String
    public var server: String
    public var transport: String
    public var capabilities: [String]
    public var requestedMuxMode: String
    public var effectiveMuxMode: String
    public var muxReceiverOrdering: String
    public var muxActiveLinks: UInt16
    public var muxFallbackReason: String
    public var p2pState: P2PState
    public var effectivePath: String { p2pState == .direct ? "direct" : "relay" }
    public var traffic: RuntimeTrafficSnapshot

    /// `monotonicMs` at which the session entered `connected`, or 0 when it is
    /// not connected. Elapsed time is `monotonicMs - connectedMonotonicMs`, so
    /// it stays correct across an app process restart.
    public var connectedMonotonicMs: UInt64
    public var lastError: RuntimeErrorSnapshot

    public init(
        generation: UInt64,
        monotonicMs: UInt64,
        phase: RuntimePhase,
        role: String = "",
        server: String = "",
        transport: String = "",
        capabilities: [String] = [],
        requestedMuxMode: String = "",
        effectiveMuxMode: String = "",
        muxReceiverOrdering: String = "",
        muxActiveLinks: UInt16 = 0,
        muxFallbackReason: String = "",
        p2pState: P2PState = .disabled,
        traffic: RuntimeTrafficSnapshot = RuntimeTrafficSnapshot(),
        connectedMonotonicMs: UInt64 = 0,
        lastError: RuntimeErrorSnapshot = RuntimeErrorSnapshot()
    ) {
        self.generation = generation
        self.monotonicMs = monotonicMs
        self.phase = phase
        self.role = role
        self.server = server
        self.transport = transport
        self.capabilities = capabilities
        self.requestedMuxMode = requestedMuxMode
        self.effectiveMuxMode = effectiveMuxMode
        self.muxReceiverOrdering = muxReceiverOrdering
        self.muxActiveLinks = muxActiveLinks
        self.muxFallbackReason = muxFallbackReason
        self.p2pState = p2pState
        self.traffic = traffic
        self.connectedMonotonicMs = connectedMonotonicMs
        self.lastError = lastError
    }

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case generation
        case monotonicMs = "monotonic_ms"
        case phase
        case role
        case server
        case transport
        case capabilities
        case requestedMuxMode = "requested_mux_mode"
        case effectiveMuxMode = "effective_mux_mode"
        case muxReceiverOrdering = "mux_receiver_ordering"
        case muxActiveLinks = "mux_active_links"
        case muxFallbackReason = "mux_fallback_reason"
        case p2pState = "p2p_state"
        case effectivePath = "effective_path"
        case traffic
        case connectedMonotonicMs = "connected_monotonic_ms"
        case lastError = "last_error"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let version = try container.decode(UInt32.self, forKey: .schemaVersion)
        guard version == Self.schemaVersion else {
            throw DecodingError.dataCorruptedError(
                forKey: .schemaVersion,
                in: container,
                debugDescription: "Unsupported runtime schema version: \(version)"
            )
        }

        generation = try container.decode(UInt64.self, forKey: .generation)
        monotonicMs = try container.decode(UInt64.self, forKey: .monotonicMs)
        phase = try container.decode(RuntimePhase.self, forKey: .phase)
        role = try container.decodeIfPresent(String.self, forKey: .role) ?? ""
        server = try container.decodeIfPresent(String.self, forKey: .server) ?? ""
        transport = try container.decodeIfPresent(String.self, forKey: .transport) ?? ""
        if container.contains(.capabilities) {
            capabilities = try container.decodeIfPresent([String].self, forKey: .capabilities) ?? []
        } else {
            capabilities = Self.bundledCapabilities
        }
        requestedMuxMode = try container.decodeIfPresent(String.self, forKey: .requestedMuxMode) ?? ""
        effectiveMuxMode = try container.decodeIfPresent(String.self, forKey: .effectiveMuxMode) ?? ""
        muxReceiverOrdering = try container.decodeIfPresent(String.self, forKey: .muxReceiverOrdering) ?? ""
        muxActiveLinks = try container.decodeIfPresent(UInt16.self, forKey: .muxActiveLinks) ?? 0
        muxFallbackReason = try container.decodeIfPresent(String.self, forKey: .muxFallbackReason) ?? ""
        p2pState = P2PState.parse(
            try container.decodeIfPresent(String.self, forKey: .p2pState) ?? "disabled"
        )
        traffic = try container.decodeIfPresent(RuntimeTrafficSnapshot.self, forKey: .traffic) ?? RuntimeTrafficSnapshot()
        connectedMonotonicMs = try container.decodeIfPresent(UInt64.self, forKey: .connectedMonotonicMs) ?? 0
        lastError = try container.decodeIfPresent(RuntimeErrorSnapshot.self, forKey: .lastError) ?? RuntimeErrorSnapshot()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(Self.schemaVersion, forKey: .schemaVersion)
        try container.encode(generation, forKey: .generation)
        try container.encode(monotonicMs, forKey: .monotonicMs)
        try container.encode(phase, forKey: .phase)
        try container.encode(role, forKey: .role)
        try container.encode(server, forKey: .server)
        try container.encode(transport, forKey: .transport)
        try container.encode(capabilities, forKey: .capabilities)
        try container.encode(requestedMuxMode, forKey: .requestedMuxMode)
        try container.encode(effectiveMuxMode, forKey: .effectiveMuxMode)
        try container.encode(muxReceiverOrdering, forKey: .muxReceiverOrdering)
        try container.encode(muxActiveLinks, forKey: .muxActiveLinks)
        try container.encode(muxFallbackReason, forKey: .muxFallbackReason)
        try container.encode(p2pState.rawValue, forKey: .p2pState)
        try container.encode(effectivePath, forKey: .effectivePath)
        try container.encode(traffic, forKey: .traffic)
        try container.encode(connectedMonotonicMs, forKey: .connectedMonotonicMs)
        try container.encode(lastError, forKey: .lastError)
    }

    public func availableMuxModes(experimental: Bool = false) -> [String] {
        ["compat", "flow", "balance", "stripe"].filter {
            capabilities.contains("mux.\($0)") && ($0 != "stripe" || experimental)
        }
    }

    public var effectiveMuxDisplayName: String {
        effectiveMuxMode == "compat" ? "Compatibility mode" : effectiveMuxMode
    }

    public var muxDiagnosticLines: [String] {
        var lines: [String] = []
        if !requestedMuxMode.isEmpty { lines.append("Requested VMUX: \(requestedMuxMode)") }
        if !effectiveMuxMode.isEmpty { lines.append("Effective VMUX: \(effectiveMuxDisplayName)") }
        if !muxFallbackReason.isEmpty { lines.append("Fallback reason: \(muxFallbackReason)") }
        return lines
    }

    public var effectivePathDisplayName: String {
        effectivePath == "direct" ? "Direct" : "Relay"
    }

    public var p2pDiagnosticLines: [String] {
        ["P2P: \(p2pState.displayName)", "Effective path: \(effectivePathDisplayName)"]
    }
}
